src/components/Verse/AdvancedCopy/utils/copyText.ts

```diff
@@ -0,0 +1,71 @@
+import clipboardCopy from 'clipboard-copy';
+
+import getTextToCopy from './getTextCopy';
+import validateRangeSelection from './validateRangeSelection';
+
+/**
+ * - Validate the range selection
+ * - Copy the text to clipboard
+ * - update the state if succeed or error
+ * - set the error message if there's an error
+ */
+
+const copyText = async ({
+  showRangeOfVerses,
+  rangeEndVerse,
+  rangeStartVerse,
+  shouldCopyFootnotes,
+  shouldCopyText,
+  translations,
+  verseKey,
+}) => {
+  // if a range is selected, we need to validate it first
+  if (showRangeOfVerses) {
+    const validationError = validateRangeSelection(rangeStartVerse, rangeEndVerse);
+    // if the validation fails
+    if (validationError) {
+      throw validationError;
+    }
+  }
+
+  const getTextBlobPromise = () =>
+    getTextToCopy({
+      rangeEndVerse,
+      rangeStartVerse,
+      shouldCopyFootnotes,
+      shouldCopyText,
+      showRangeOfVerses,
+      translations,
+      verseKey,
+    }).then((text) => new Blob([text], { type: 'text/plain' }));
+
+  const textBlobPromise = getTextBlobPromise();
+
+  /**
+   * Safari supports `text/plain` : promise
+   * Chrome doesn't support it.
+   *
+   * Safari needs `navigator.clipboard.write` to be called immediately without waiting
+   * Chrome can wait
+   *
+   * So, for safari we call it navigator.clipboard immediately and give it a promise
+   * for chrome we wait for the promise to resolve, the wen call navigator.clipboard.write
+   */
+
+  try {
+    copyOnSafari(textBlobPromise);
```

Comment:
We need to move this to a global util function that can be re-used.

Reply:
Also we should rename `copyOnSafari` since the way we are using is not only Safari specific but rather supports multiple browsers